src/on_shutdown.c

```diff
@@ -0,0 +1,182 @@
+#include "on_shutdown.h"
+
+#include <stdlib.h>
+#include <small/rlist.h>
+#include <trivia/util.h>
+#include <fiber.h>
+#include <libev/ev.h>
+
+struct on_shutdown_trigger {
+	/*
+	 * rlist entry
+	 */
+	struct rlist entry;
+	/*
+	 * Shutdown handler arg
+	 */
+	void *arg;
+	/*
+	 * Shutdown handler
+	 */
+	void (*handler)(void *);
+	/*
+	 * Async event set when module main thread finished
+	 */
+	struct ev_async on_shutdown_async;
+	/*
+	 * Module shutdown fiber
+	 */
+	struct fiber *on_shutdown_fiber;
+	/*
+	 * Main loop
+	 */
+	struct ev_loop *loop;
+};
+
+/*
+ * Shutdown trigger list, not need mutex, because access is
+ * avaliable only from tx thread
+ */
+static RLIST_HEAD(on_shutdown_trigger_list);
+
+static void
+on_shutdown_fiber_wakeup(MAYBE_UNUSED struct ev_loop *loop, ev_async *ev, MAYBE_UNUSED int revents)
+{
+	struct on_shutdown_trigger *on_shutdown_trigger = ev->data;
+	fiber_wakeup(on_shutdown_trigger->on_shutdown_fiber);
+}
+
+static int
+create_new_on_shutdown_trigger(void *arg, void (*handler)(void *), void **opaque)
```

Comment:
I'd rename to:
(on_)module_shutdown_trigger_create (which is consistent with trigger_create());
(on_)module_shutdown_trigger_new (which is consistent with txn_alter_trigger_new()).

Reply:
Also I'd swap the order of arguments: I'd pass function ptr first, and then its argument.

Reply:
fix